lib/charms/mongodb/v0/shards_interface.py

```diff
@@ -53,27 +73,16 @@ def __init__(
         self.framework.observe(
             charm.on[self.relation_name].relation_changed, self._on_relation_event
         )
+        self.framework.observe(
+            charm.on[self.relation_name].relation_broken, self._on_relation_event
+        )
 
-        # TODO Follow up PR, handle rotating passwords
+        # TODO Future PR: handle self healing when all IP addresses of a shard changes and we have
+        # to manually update mongos
 
     def _on_relation_joined(self, event):
         """Handles providing shards with secrets and adding shards to the config server."""
-        if self.charm.is_role(Config.Role.REPLICATION):
-            self.charm.unit.status = BlockedStatus("role replication does not support sharding")
-            logger.error("sharding interface not supported with config role=replication")
-            return
-
-        if not self.charm.is_role(Config.Role.CONFIG_SERVER):
-            logger.info(
-                "skipping relation joined event ShardingRequirer is only be executed by config-server"
-            )
-            return
-
-        if not self.charm.unit.is_leader():
-            return
-
-        if not self.charm.db_initialised:
-            event.defer()
+        if not self.pass_hook_checks(event):
             return
```

Comment:
Should we defer `if not self.charm.db_initialised`?

Reply:
Yes we should defer, (and [we do defer in this case, in the function `pass_hook_checks `](https://github.com/canonical/mongodb-operator/blob/e8d7167762a262f6bcaf3a8c3c5460a46b720c86/lib/charms/mongodb/v0/shards_interface.py#L122-L126)) 

Reasoning to defer is that we are not guaranteed to receive another `relation_joined` event, so we should keep it on the event queue until we can process the event